Merge conflicts exercise (#92)
rstudio-education/datascience-box: course-materials/lab-instructions/lab-06/lab-06-ugly-charts.Rmd

This is the second week you're working in teams, so we're going to make things a little more interesting and let all of you make changes and push those changes to your team repository. Sometimes things will go swimmingly, and sometimes you'll run into merge conflicts. So our first task today is to walk you through a merge conflict!

## Merges and merge conflicts

- Pushing to a repo replaces the code on GitHub with the code you have on your computer. 
- If a collaborator has made a change to your repo on GitHub that you haven't incorporated into your local work, GitHub will stop you from pushing to the repo because this could overwrite your collaborator's work!
- So you need to explicitly "merge" your collaborator's work before you can push.
- If your and your collaborator's changes are in different files or in different parts of the same file, git merges the work for you automatically when you *pull*.

> - If you both changed the same part of a file, git will produce a **merge conflict** because it doesn't know how which change to you want to keep and which change you want to overwrite.

Reviewer: ```suggestion
- If you both changed the same part of a file, git will produce a **merge conflict** because it doesn't know how which change you want to keep and which change you want to overwrite.
```

  
Git will put conflict markers in your code that look like:

```
<<<<<<< HEAD
See also: [re docs](https://docs.python.org/3/library/re.html)  
=======
See also [re docs](https://docs.python.org/3/library/re.html)  
>>>>>>> master
```

The `===`s separate *your* changes (top) from *their* changes (bottom)

Your job is to *reconcile* the changes: edit the file so that it incorporates
the best of both versions and delete the `<<<`, `===`, and `>>>` lines. Then
Stage and Commit the result.

### Setup

- Clone the repo and open the .Rmd file.
- Assign the numbers 1, 2, 3, and 4 to each of the team members. Some people will need to have multiple numbers.

### Let's cause a merge conflict!

Our goal is to see two different types of merges: first we'll see one that git
can't figure out on its own (a "merge conflict"), then one that it can figure out.

Doing this will require some tight choreography, so pay attention.

Take turns in completing the exercise, only one member at a time. **Others should just watch, not doing anything on their own checkouts (including pull)** until they are instructed to.

> - **Before starting**: everyone should have the repo checked out and know which role number(s) they are.

Reviewer: ```suggestion
- **Before starting**: everyone should have the repo cloned and know which role number(s) they are.
```

    all have check marks, not filled-in boxes.
  * Commit and push.  

`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.

- **Role 3:**

  * Add a label to the first code chunk
  * Knit, commit, push. You should get an error.
  * Pull. No merge conflicts should occur, but you should see a message about merging.
  * Now push.

`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.

- **Role 4:**

  * Add a different label to the first code chunk.
  * Knit, commit, push. You should get an error.
  * Pull. Take a look at the document with the merge conflict. Clear the merge conflict by choosing the correct/preferred change. Commit, and push.  
`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.  

- **All members:** Pull, and observe the changes in your document.

### Tips for collaborating via GitHub

- Always pull first before you start working.
- Resolve a merge conflict (commit and push) *before* continuing your work. Never do
  new work while resolving a merge conflict.
- Knit, Commit, and Push often to minimize merge conflicts and/or to make merge conflicts easier to resolve.

> - If you find yourself in a situation that is difficult to resolve, ask questions ASAP. Don't let it linger and get bigger.
> 
> (Optional further reading: [Merge conflicts in the classroom](https://github.blog/2018-08-22-merge-conflicts-in-the-classroom/).)

Reviewer: ```suggestion
- If you find yourself in a situation that is difficult to resolve, ask questions ASAP. Don't let it linger and get bigger.
```

This is the second week you're working in teams, so we're going to make things a little more interesting and let all of you make changes and push those changes to your team repository. Sometimes things will go swimmingly, and sometimes you'll run into merge conflicts. So our first task today is to walk you through a merge conflict!

## Merges and merge conflicts

> - Pushing a repo replaces the code on GitHub with the code you have on your computer. 

Reviewer: ```suggestion
- Pushing to a repo replaces the code on GitHub with the code you have on your computer. 
```

Reviewer: Also, the concept of "your computer" is a bit fuzzy when working on the cloud but I think this is ok / not confusing even though technically not accurate. I go back and forth about accuracy around this.

Author: Maybe the missing concept is that you have a working repo (on your computer or on RStudio or whatever) that you can commit to directly but you update the shared repo on GitHub by pushing.

> - If someone else has made a change that you haven't incorporated into your repo, GitHub will stop you from pushing because it would lose their work!

Reviewer: ```suggestion
- If a collaborator has made a change to your repo on GitHub that you haven't incorporated into your local work, GitHub will stop you from pushing to the repo because this could overwrite your collaborator's work!
```

> - So you need to explicitly "merge" your collaborators' work before you pull.

Reviewer 2: Should say "push" in this line, not pull, right?

Author: Good catch.

> - So you need to explicitly "merge" your collaborators' work before you push.

Reviewer: ```suggestion
- So you need to explicitly "merge" your collaborator's work before you can push.
```

> - If the changes are in different files, or different parts of the same file,
>   git merges the work for you automatically when you *pull*.

Reviewer: ```suggestion
- If your and your collaborator's changes are in different files or in different parts of the same file, git merges the work for you automatically when you *pull*.
```

> - If you both changed the same part of a file, git will produce a **merge conflict**
>   because it doesn't know how which change to take.

Reviewer: ```suggestion
- If you both changed the same part of a file, git will produce a **merge conflict** because it doesn't know how which change to you want to keep and which change you want to overwrite.
```

Author: Typo (“to you”)

  
Git will put conflict markers in your code that look like:

```

> <<<<<<< HEAD
> See also: [re docs](https://docs.python.org/3/library/re.html)  
> =======
> See also [re docs](https://docs.python.org/3/library/re.html)  
> >>>>>>> master

Reviewer: ```suggestion
<<<<<<< HEAD
See also: [dplyr documentation](https://dplyr.tidyverse.org/)  
=======
See also [ggplot2 documentation](https://ggplot2.tidyverse.org/)  
>>>>>>> some1alpha2numeric3string4
```

Reviewer: I made the difference more obvious and also R related (just so we reduce the number of unfamiliar things they see within this activity).

Also, it shouldn't be `master` but a hash for the last commit so made an "informative" text string but I wonder if an actual hash is better...

```

> The `===`s separate *your* changes (top) from *their* changes (bottom)

Reviewer: ```suggestion
The `===`s separate *your* changes (top) from *their* changes (bottom). 
Note that on top you see the word `HEAD`, which indicates that these are your changes. 
And at the bottom you see `some1alpha2numeric3string4` (well, it probably looks more like `28e7b2ceb39972085a0860892062810fb812a08f`). 
This is the *hash* (a unique identifier) of the commit your collaborator made with the conflicting change. 
```

Your job is to *reconcile* the changes: edit the file so that it incorporates
the best of both versions and delete the `<<<`, `===`, and `>>>` lines. Then
Stage and Commit the result.

### Setup

- Clone the repo and open the .Rmd file.

> - Assign the numbers 1, 2, 3, and 4 to each of the team members. Some people will need to have multiple numbers.

Reviewer: ```suggestion
- Assign the numbers 1, 2, 3, and 4 to each of the team members. If your team has less than 4 people, some people will need to have multiple numbers. If your team has more than 4 people, some people will need to share some numbers.
```

### Let's cause a merge conflict!

Our goal is to see two different types of merges: first we'll see one that git
can't figure out on its own (a "merge conflict"), then one that it can figure out.

Doing this will require some tight choreography, so pay attention.

> Take turns in completing the exercise, only one member at a time. **Others should just watch, not doing anything on their own checkouts (including pull)** until they are instructed to.

Reviewer: Do you teach the term "checkout"? Could we say this without that term?

Author: That was accidentally a new concept, good catch. The private/temporary/work repo concept i mentioned above would handle this.

> - **Before starting**: everyone should have the repo checked out and know which role number(s) they are.

Reviewer: I'm not sure what "repo checked out" means here?

Author: I meant cloned.

> Take turns in completing the exercise, only one member at a time. **Others should just watch, not doing anything on their own checkouts (including pull)** until they are instructed to.

Reviewer: ```suggestion
Take turns in completing the exercise, only one member at a time. **Others should just watch, not doing anything on their own projects (this includes not even pulling changes!)** until they are instructed to.
```

- **Role 1:**

  * Change the team name to your actual team name.
  * Knit, commit, push.  

`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.

- **Role 2:**

  * Change the team name to some other word.
  * Knit, commit, push. You should get an error.
  * Pull. Take a look at the document with the merge conflict.

>   * Clear the merge conflict by choosing the correct/preferred change.

Reviewer: ```suggestion
  * Clear the merge conflict by editing the document to choose the correct/preferred change.
```

- **Role 3:**

  * Add a label to the first code chunk
  * Knit, commit, push. You should get an error.
  * Pull. No merge conflicts should occur, but you should see a message about merging.
  * Now push.

`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.

- **Role 4:**

  * Add a different label to the first code chunk.
  * Knit, commit, push. You should get an error.
  * Pull. Take a look at the document with the merge conflict. Clear the merge conflict by choosing the correct/preferred change. Commit, and push.  
`r emo::ji("stop_sign")` Wait for instructions before moving on to the next step.  

- **All members:** Pull, and observe the changes in your document.

### Tips for collaborating via GitHub

- Always pull first before you start working.
- Resolve a merge conflict (commit and push) *before* continuing your work. Never do
  new work while resolving a merge conflict.
- Knit, Commit, and Push often to minimize merge conflicts and/or to make merge conflicts easier to resolve.
- If you find yourself in a situation that is difficult to resolve, ask questions ASAP. Don't let it linger and get bigger.

> (Optional further reading: [Merge conflicts in the classroom](https://github.blog/2018-08-22-merge-conflicts-in-the-classroom/).)

Reviewer: As much as I love Vanessa's writing on this topic, I hesitate to include this here because the workflow is different than what the curriculum teaches. I find that mixing and matching Terminal instructions with RStudio Git pane instructions confuses students who are new to computing and version control. So I'd suggest taking this out. Unfortunately Happy Git with R also addresses this at a level higher than what's appropriate for this curriculum, IMO.

Author: Agreed.
